src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -27,36 +27,57 @@ class CreateZuoraSubscription(servicesProvider: ServiceProvider = ServiceProvide
-      case None => subscribe(state, requestInfo, services)
-    }
-  } yield thankYouState
+  ): FutureHandlerResult = {
+    val subscribeItem = buildSubscribeItem(state, services.promotionService)
+    for {
+      identityId <- Future.fromTry(IdentityId(state.user.id))
+      maybeDomainSubscription <- GetRecurringSubscription(services.zuoraService, state.requestId, identityId, state.product.billingPeriod)
+      previewPaymentSchedule <- PreviewPaymentSchedule(subscribeItem, state.product.billingPeriod, services, checkSingleResponse)
+      thankYouState <- maybeDomainSubscription match {
+        case Some(domainSubscription) => skipSubscribe(state, requestInfo, domainSubscription)
+        case None => subscribe(state, subscribeItem, previewPaymentSchedule, requestInfo, services)
+      }
+    } yield thankYouState
+  }
 
   def skipSubscribe(state: CreateZuoraSubscriptionState, requestInfo: RequestInfo, subscription: DomainSubscription): FutureHandlerResult = {
     val message = "Skipping subscribe for user because they are already an active contributor"
     SafeLogger.info(message)
-    FutureHandlerResult(getEmailState(state, subscription.accountNumber, subscription.subscriptionNumber), requestInfo.appendMessage(message))
+    FutureHandlerResult(
+      getEmailState(state, subscription.accountNumber, subscription.subscriptionNumber, PaymentSchedule(Nil)), requestInfo.appendMessage(message)
+    )
   }
 
   def singleSubscribe(
     multiSubscribe: SubscribeRequest => Future[List[SubscribeResponseAccount]]
   ): SubscribeItem => Future[SubscribeResponseAccount] = { subscribeItem =>
-    multiSubscribe(SubscribeRequest(List(subscribeItem))) flatMap {
+    checkSingleResponse(multiSubscribe(SubscribeRequest(List(subscribeItem))))
+  }
+
+  def checkSingleResponse[ResponseItem](response: Future[List[ResponseItem]]): Future[ResponseItem] = {
+    response.flatMap {
       case result :: Nil => Future.successful(result)
       case results => Future.failed(new RuntimeException(s"didn't get a single response item, got: $results"))
     }
   }
 
-  def subscribe(state: CreateZuoraSubscriptionState, requestInfo: RequestInfo, services: Services): FutureHandlerResult =
-    singleSubscribe(services.zuoraService.subscribe)(buildSubscribeRequest(state, services.promotionService))
+  def subscribe(
+    state: CreateZuoraSubscriptionState,
+    subscribeItem: SubscribeItem,
+    previewedPaymentSchedule: PaymentSchedule,
+    requestInfo: RequestInfo,
+    services: Services
+  ): FutureHandlerResult =
+    singleSubscribe(services.zuoraService.subscribe)(subscribeItem)
       .map(response =>
-        HandlerResult(getEmailState(state, response.domainAccountNumber, response.domainSubscriptionNumber), requestInfo))
+        HandlerResult(getEmailState(state, response.domainAccountNumber, response.domainSubscriptionNumber, previewedPaymentSchedule), requestInfo))
```

Comment:
it doesn't seem necessary to pass this in only to pass it straight on to the HandlerResult.  This also applies to requestInfo which is existing.

Reply:
We could do something like this instead:

```
thankYouState <- maybeDomainSubscription match {
  case Some(domainSubscription) => skipSubscribe(state, requestInfo, previewPaymentSchedule, domainSubscription)
  case None => subscribe(state, subscribeItem, services).map(response => toHandlerResult(state, response, previewPaymentSchedule, requestInfo))
}
```

Do you think that's preferable?

Reply:
yes that sounds better, then we reduced the number of parameters from 5 to 3.  Also your way looks more normal that what I imagined in my head! 💯